#-*- mode:snakemake -*-

extra_conf_fn = srcdir('alevin.config')
if os.path.exists(extra_conf_fn):
    with open(extra_conf_fn) as fh:
        c  = yaml.load(fh) or {}
        update_config2(config, c)

#config
FASTQ_DIR = config.get('fastq_dir', 'data/raw/fastq')
AVN_INTERIM = join(INTERIM_DIR, 'alevin')
AVN_PROCESSED = join(PROCESSED_DIR, 'alevin')
REF = config['quant']['reference']

def get_fastq_R1(wildcards):
    R1 = config['samples'][wildcards.sample]['R1'].split(',')
    R1 = [join(FASTQ_DIR, fn) for fn in R1]
    return ' '.join(R1)

def get_fastq_R2(wildcards):
    R2 = config['samples'][wildcards.sample]['R2'].split(',')
    R2 = [join(FASTQ_DIR, fn) for fn in R2]
    return ' '.join(R2)

rule alevin_10xgenomics_whitelist:
    input:
        join(INTERIM_DIR, '10xgenomics', REF, '{sample}', 'outs', 'filtered_feature_bc_matrix', 'barcodes.tsv.gz')
    output:
        join(AVN_INTERIM, 'whitelists', '{sample}', 'whitelist.txt')
    shell:
        'zcat {input} > {output}'
                     
rule alevin_quasi:
    input:
        R1 = get_fastq_R1,
        R2 = get_fastq_R2,
        ref = join(EXT_DIR, REF, 'salmon', 'refInfo.json'),
        t2g = join(EXT_DIR, REF, 'genes', 'txp2gene.tsv')
    params:
        args = '-l ISR --chromiumV3',
        output = join(AVN_PROCESSED, '{sample}'),
        ref = join(EXT_DIR, REF, 'salmon')
    singularity:
        'docker://flatberg/alevin:dev0.12'
    output:
        join(AVN_PROCESSED, '{sample}', 'alevin', 'quants_mat.gz')
    shell:
        'salmon alevin '
        '{params.args} '
        '-i {params.ref} '
        '--tgMap {input.t2g} '
        '-1 {input.R1} '
        '-2 {input.R2} '
        '-p {threads} '
        '-o {params.output} '
