#-*- mode:snakemake -*-

extra_conf_fn = srcdir('star.config')
if os.path.exists(extra_conf_fn):
    with open(extra_conf_fn) as fh:
        c  = yaml.load(fh) or {}
        update_config2(config, c)

#config
ORG = config.get('organism', 'homo_sapiens')
FASTQ_DIR = config.get('fastq_dir', 'data/raw/fastq')
STAR_PROCESSED = join(QUANT_PROCESSED, 'star')
STAR_INTERIM = join(QUANT_INTERIM, 'star')
STAR_CONF = config['quant'].get('star', {})
REF = config['quant']['reference']

def starsolo_R1(wildcards):
    return ','.join(get_filtered_fastq_R1(wildcards))

def starsolo_R2(wildcards):
    return ','.join(get_filtered_fastq_R2(wildcards))

rule starsolo:
    input:
        R1 = get_filtered_fastq_R1,
        R2 = get_filtered_fastq_R2,
        transcriptome = join(ENSEMBL_EXT, 'star', 'SA'),
        whitelist = rules.txgenomics_whitelist.output
    params:
        outdir = join(STAR_INTERIM, '{sample}') + '/',
        genome_dir = join(ENSEMBL_EXT, 'star'),
        cb_len = STAR_CONF['cb_len'],
        umi_len = STAR_CONF['umi_len'],
        R1 = starsolo_R1,
        R2 = starsolo_R2
    threads:
        48
    output:
        barcodes = join(STAR_INTERIM, '{sample}', 'Solo.out', 'barcodes.tsv'),
        gene_stats = join(STAR_INTERIM, '{sample}', 'Solo.out', 'Gene.stats'),
        genes = join(STAR_INTERIM, '{sample}', 'Solo.out', 'genes.tsv'),
        feature_mtx = join(STAR_INTERIM, '{sample}', 'Solo.out', 'matrix.mtx'),
    singularity:
        STAR_CONF['docker']
    benchmark:
        'benchmark/starsolo/{sample}-starsolo.txt'
    shell:
        'STAR --soloType Droplet '
        '--soloCBwhitelist {input.whitelist} '
        '--readFilesIn {params.R2} {params.R1} '
        '--readFilesCommand zcat '
        '--genomeDir {params.genome_dir} '
        '--outFileNamePrefix {params.outdir} '
        '--runThreadN {threads} '
        '--soloCBlen {params.cb_len} '
        '--soloUMIlen {params.umi_len} '
        '--genomeLoad LoadAndKeep '

rule star_scanpy:
    input:
        join(STAR_INTERIM, '{sample}', 'Solo.out', 'matrix.mtx')
    params:
        script = srcdir('scripts/cellranger_scanpy.py'),
        input = join(STAR_INTERIM, '{sample}', 'Solo.out')
    output:
        join(STAR_PROCESSED, 'scanpy', '{sample}', '{sample}.h5ad')
    singularity:
        'docker://gcfntnu/scanpy:1.3.7'
    shell:
        'python {params.script} -i {params.input} -o {output} '
 
